WIP integrate averager into mnefun

ktavabi opened the pull request:
@jjnurminen here I am using `mne.combine_evoked` as per @Eric89GXL to weigh trials with cHPI SNR.

ktavabi committed:
WIP integrate averager into mnefun via mne.combine_evoked

ktavabi commented on mnefun/_mnefun.py:
```diff
@@ -1409,8 +1414,16 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
+                if p.weighting:
+                    weights = _chpi_snr_epochs(e[name])
+                    weigh_these = [trl for trl in e[name].iter_evoked()]
+                    evoked = combine_evoked(weigh_these, weights=weights)
```
@Eric89GXL here `combine_evoked`  does produce an evoked file with expected time course. However for some reason though:

1. I cannot visualize evoked data with `mne_browse_raw`
2. comments string in resulting evoked is overwritten by what I think is data value * weight e.g., 799550266427926038708224.000 * 1 + 759231199089828604936192.000 * 1 + 893204949043000480104448.000 * 1 + 1120020057860884452605952.000 * 1 + ...
3. NAVE is always set to 1
4. data seems to be dramatically scaled down

![image](https://cloud.githubusercontent.com/assets/4381290/21333194/f3e63468-c602-11e6-85a4-bdd43ea5d783.png)
White is ERF from All condition for funloc subj 02 and blue/cyan is the weighted 

Am I using this method right?

larsoner commented:
>
>
>    1. I cannot visualize evoked data with mne_browse_raw
>
>
It is probably maxshield'ed data. Did you try --allowmaxshield? Even then
it might not support evoked w/MS. But MNE-Python does.

>
>    1. comments string in resulting evoked is overwritten by what I think
>    is data value * weight e.g., 799550266427926038708224.000 * 1 +
>    759231199089828604936192.000 * 1 + 893204949043000480104448.000 * 1 +
>    1120020057860884452605952.000 * 1 + ...
>
> Yes you should probably set it to something else.

>
>    1. NAVE is always set to 1
>
> This is because of huge weights

>
>    1. data seems to be dramatically scaled down
>
> Try normalizing the weights before passing them to combine_evoked,
normalization (or possibly some other scale-factor, but probably
normalization) will fix this and the `nave` issue.

ktavabi commented:
> It is probably maxshield'ed data. Did you try --allowmaxshield? Even then
> it might not support evoked w/MS. But MNE-Python does.

No, I am pretty sure this is not a maxshield thing; these are single epochs from `iter_evoked` in the mnefun pipe. What is MS?

> > Try normalizing the weights before passing them to combine_evoked,
> normalization (or possibly some other scale-factor, but probably
> normalization) will fix this and the `nave` issue.

So with `nweights = weights / np.linalg.norm(weights)` as arg to `combine_evoked` the scaling is now the otherway around with weighted >> unweighted amplitudes. I don't understand why rescaling is necessary at all, what exactly is happening to the data in `combine_evoked`?

larsoner commented:
MaxShield

It is possible to have MaxShield evoked data depending on raising
conditions and whether or not Maxwell filtering was applied

ktavabi commented:
Not sure WYM by raising conditions, but I can now load the evoked DS in browse_raw after applying `nweights` from above and setting the comments field to some string type.

larsoner commented:
Sorry on mobile, I meant recording conditions

Did normalizing weights fix everything then?

ktavabi commented:
> Did normalizing weights fix everything then?

From above I tried (unit) normalization with `nweights = weights / np.linalg.norm(weights)` as arg to `combine_evoked` but now the weighted amplitudes are dwarfing the unweighted amplitudes. I looked at the `combine_evoked` code and it's only multiplying data by weights. So I think I will need to scale the data after applying the weights.

larsoner commented:
The weights are all positive, correct? Have you tried `weights /= weights.sum()`? You shouldn't need to scale the data after applying the weights, you should only need to change the weights. Multiplying the weights by 10 or dividing them by 10 before passing them to `combine_evoked` should scale the data different ways.

jjnurminen commented:
To get correct results for any arbitrary weights, you should do `weights *= (number of epochs / sum of weights)` since subsequent averaging operations will typically sum and divide by the number of epochs.

ktavabi commented:
Correct weights are all positive. Seems I chose the wrong norm `weights /= weights.sum()` works.

![image](https://cloud.githubusercontent.com/assets/4381290/21360005/ff9861d8-c692-11e6-941a-f99eb28e6c0f.png)

I will test `weights *= (number of epochs / sum of weights)` I suspect it will be the same.

jjnurminen commented:
Huh? I don't think it's gonna be the same, unless you have 1 epoch :)
The average will look reasonable, but the magnitude will be wrong (too small)
A simple illustration (w=weights, x=data):
```

x = np.array([1,2,3])
w = np.array([1/3.,1/3.,1/3.])
np.mean(x)  # 2
np.mean(x*w / np.sum(w))  # 0.6667 :(
np.mean(len(x) * x*w / np.sum(w))  # 2.0 :)

```

jjnurminen commented:
~~Actually I may be misleading you here, since I'm not sure what kind of magic `combine_evoked` does under the hood.~~
`combine_evoked` just seems to apply the raw weights, so they should be scaled as above.
It's easy to test by setting all weights to 1 and seeing that the result is exactly the same as running `epochs.average()`

ktavabi commented:
@jjnurminen you are right. But `weights *= (nepochs / weights.sum())` as arg to `combine_evoked` yields inaccurate number of effective averages for conditions i.e, `nave = 1`, b/c `combine_evoked` computes number of averages as the reciprocal of the variance of a weighted sample mean. It looks like that works out to 1 for all conditions when the weights are excessive. @Eric89GXL make sense?

On the other hand `weights /= weights.sum()` yields the correct number of effective averages needed for scaling data covariance later during source analysis among different conditions. 

Anyways, I am going to push so you guys can review.

jjnurminen commented:
To me it looks like `combine_evokeds` is trying to do something too smart. In my opinion `nave` should just reflect the new total of epochs, otherwise it's very misleading.

ktavabi committed:
feed normed weights to mne.combine_evoked

ktavabi committed:
cleaned up imports

ktavabi commented:
After 10 years I don't really have any opinions! I just do what people say Matti Hamalainen said to do, and this business of the "effective number of averages" is one of those things :smile: In all seriousness tho the recent commits result in `nave` as the new total number of epochs going into each (condition) ERF making downstream between condition comparisons accurate.

jjnurminen commented:
That's wise indeed. As an ancient Zen text says:

"If you would clearly see the truth, discard opinions pro and con.
To founder in dislike and like is nothing but the mind's disease."

Anyway, sorry for the noise, will review ASAP :)

ktavabi commented:
@jjnurminen :beer:

larsoner commented:
Sorry @Eric89GXL, I tried to quote but accidentally edited this comment of yours instead of posting a new one. (Didn't know that this was even possible...)

larsoner commented:
To see why something smart needs to be done, I can do a weighted summation of 1000 epochs where weights are the delta function `[1, 0, 0, ...]`. Clearly setting `nave=1000` is wrong here as it should be `nave=1`. The MNE code will do the right thing in this case, and other constructed cases, IIRC.

ktavabi commented:
> There has been a long discussion with several people of nave calculation on MNE-Python

Where is this?

larsoner commented:
https://github.com/mne-tools/mne-python/pull/3363
https://github.com/mne-tools/mne-python/pull/3400

ktavabi commented:
thanks

larsoner commented:
FYI if you keep looking by searching "nave" you find more 

https://github.com/mne-tools/mne-python/issues/1634
https://github.com/mne-tools/mne-python/pull/461

ktavabi commented:
My search turned up quite a bit and I found some older ones mainly about robust averaging. Missed these more recent ones.

jjnurminen commented:
>To see why something smart needs to be done, I can do a weighted summation of 1000 epochs where weights are the delta function [1, 0, 0, ...]

That's a pathological case (IMO). 

>interested in reading why it's done that way. The short story is that if you want `nave` to be meaningful in the sense of how to scale source estimates like dSPM based on an effective number of epochs, it does 

I get that, but then the name `nave` is not accurate any longer. It's more like a virtual number of averages.
Anyway, I'm just whining. Probably the practical implications are not that big. `nave` is not written out to fiff files, right?

>`weights /= weights.sum()` is the right thing here because under the hood `combine_evokeds` just does a weighted summation, not weighted average (otherwise the suggested `n_epochs * weights.sum()` would be the correct normalization).

Aha, I see. But it does do averaging when you specify `weights="equal"`, right?

jjnurminen commented on mnefun/_mnefun.py:
```diff
@@ -242,6 +242,9 @@ class Params(Frozen):
         Default is center of sphere fit to digitized head points.
     weighting : bool | False
         Average epochs weighted by the signal-to-noise ratio of continuous HPI.
+        Defualts to 5 harmonics to model HPI frequency harmonics for each HPI
```
The harmonics are not for HPI, but for the line frequency.

larsoner commented:
On Dec 21, 2016 3:36 AM, "Jussi Nurminen" <notifications@github.com> wrote:

To see why something smart needs to be done, I can do a weighted summation
of 1000 epochs where >weights are the delta function [1, 0, 0, ...]

That's a pathological case (IMO).

Yes it's unlikely to happen, but it would be nice if the calculation
exhibited the behavior of giving nave=1 in this case. And from this you can
hopefully see how it generalizes to real world cases that could arise (few
epochs with high weight, many epochs with negligible weights), including
with your cHPI scheme.

interested in reading why it's done that way. The short story is that if
you want nave to be meaningful in >the sense of how to scale source
estimates like dSPM based on an effective number of epochs, it does

I get that, but then the name nave is not accurate any longer. It's more
like a virtual number of averages.

It is accurate in terms of what is happening in the standard Gaussian noise
model.

Anyway, I'm just whining. Probably the practical implications are not that
big. nave is not written out to fiff files, right?

It is written out.

weights /= weights.sum() is the right thing here because under the hood
combine_evokeds just does a >weighted summation, not weighted average
(otherwise the suggested n_epochs * weights.sum() would >be the correct
normalization).

Aha, I see. But it does do averaging when you specify weights="equal",
right?

Yes, and it does so using the same nave calculation so it generalizes.

jjnurminen commented:
[nave]
>It is written out

In that case, is there a chance that subsequent users of the fiff file get confused?
For example, I'm not sure how e.g. Elekta software uses the `nave `value - I could try to check. Or did you already go through this in the mne-python discussions?

jjnurminen commented:
Well it looks like Matti Hämäläinen was involved in the discussions and he is intimately familiar with Elekta software. So probably it's ok.

ktavabi commented:
Guys @jjnurminen @Eric89GXL are we in agreement with the useage of `mne.combine_evoked` for integrating chpi weights into averaging?

jjnurminen commented:
@kambysese If Eric thinks that using `combine_evokeds` is a better way, that's fine by me

larsoner commented:
Yes please use `combine_evokeds`

ktavabi commented:
I did use combined_evoked in last commit. I was checking to see if either
of you had any comments on the implementation, I.e, push forward with the
review process.

------------------------------------------------------------
Kambiz Tavabi PhD
Institute for Learning & Brain Sciences
1715 Columbia Road N
Portage Bay Building
Box 357988
University of Washington
Seattle, WA 98195-7988
Tel: 206-685-6173
------------------------------------------------------------

*Confidentiality Notice: Because email is not secure, please be aware that
we cannot guarantee the confidentiality of information sent by email. If
you are not the intended recipient of this email, please notify the sender
by reply email, and then destroy all copies of the message and any
attachments.*

On January 11, 2017 at 6:05:52 AM, Eric Larson (notifications@github.com)
wrote:

> Yes please use combine_evokeds
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jjnurminen/mnefun/pull/2#issuecomment-271876561>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AELaan1S8Muyouhb0b6sUsa31IPw4J-zks5rROHAgaJpZM4LRTvF>
> .
>

larsoner commented on examples/funloc/analysis_fun.py:
```diff
@@ -124,7 +125,7 @@
     plot_psd=True,      # Plot raw data power spectra
     write_epochs=True,  # Write epochs to disk
-    gen_covs=True,      # Generate covariances
+    gen_covs=False,      # Generate covariances
```
revert

larsoner commented on mnefun/_mnefun.py:
```diff
@@ -239,6 +240,11 @@ class Params(Frozen):
     sss_origin : array-like, shape (3,) | str
         Origin of internal and external multipolar moment space in meters.
         Default is center of sphere fit to digitized head points.
+    weighting : bool | False
```
`bool`

But this isn't a great future-compatible mode. I'd probably make it `str | None` and allow only `'chpi'` as a string option for now. And maybe it should be called `average_weighting` since it's only used for creating evoked

larsoner commented on mnefun/_mnefun.py:
```diff
@@ -1409,8 +1417,19 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
+                    #TODO give access to _chpi_snr_epochs args?
+                    weights = _chpi_snr_epochs(e[name])
+                    weights /= weights.sum()  # normalize weights
+                    weigh_these = [trl for trl in e[name].iter_evoked()]
```
`weigh_these = list(e['name'].iter_evoked())`

larsoner commented on mnefun/_mnefun.py:
```diff
@@ -1409,8 +1417,19 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
-                evokeds.append(e[name].average())
-                evokeds.append(e[name].standard_error())
+                if p.weighting:
+                    #TODO give access to _chpi_snr_epochs args?
```
We probably shouldn't need to, assuming the defaults there work properly.

larsoner reviewed (commented):
Otherwise LGTM

ktavabi committed:
Address review comments.

ktavabi commented:
@jjnurminen ready for merge from my end cc @Eric89GXL

ktavabi committed:
ignore ipynb checkpoints

ktavabi committed:
Merge branch 'chpi_averager' of https://github.com/kambysese/mnefun into chpi_averager

ktavabi commented on mnefun/_mnefun.py:
```diff
@@ -1417,18 +1418,17 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
-                    weigh_these = [trl for trl in e[name].iter_evoked()]
+                    weigh_these = list(e[name].iter_evoked())
                     evoked = combine_evoked(weigh_these, weights=weights)
                     evoked.comment = name
```
@Eric89GXL ~~can you think of any reason why I would be losing epochs here?
EDIT `evoked.nave` is set to 130 although the number of epochs for condition in question is 139 (All good). Why is that?~~ I see what's going on; nave is 1/the variance of the weighted sample mean.

I am wondering whether it would not be safer to search and load `epo.fif` file rather than re-do epoching in order to prevent accidental changes in epoching being introduced into weighted averaging. WDYT?

larsoner replied:
That shouldn't work -- the `epo.fif` that we save has the cHPI data filtered out already

ktavabi replied:
Understood, but if we decide to go forward with this particular implementation I don't think it is safe to re-run the epoching prior to weighting. I suspect in most use cases users will want to weight epoched data they have on hand.

larsoner replied:
> I don't think it is safe to re-run the epoching prior to weighting

What are you worried about in this regard?

ktavabi commented on mnefun/_mnefun.py:
```diff
@@ -1417,18 +1418,17 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
-                if p.weighting:
-                    #TODO give access to _chpi_snr_epochs args?
+                if p.average_weighting == 'chpi':
                     weights = _chpi_snr_epochs(e[name])
```
@jjnurminen the input to `_chpi_snr_epochs` here is epochs that were created after movement compensation and additional processing. Does this result in accurate weights i.e., same as weights computed from raw file prior to movement compensation? That is not my understanding

jjnurminen replied:
@kambysese The important thing is that the cHPI signals are unaltered in the data that is used for computing the weights. I'm not yet 100% sure what's happening here, but if the data is movement compensated, the cHPI signals may indeed be distorted (or totally gone).

ktavabi commented:
@jjnurminen My feeling here is that it will be difficult to get much progress on a manuscript for this project if:
(1) `chpi_weighted_average` script works with `epo.fif` file as arg
OR
(2) `_chpi_snr_epochs` function computes weights for trials prior to movement compensation e.g., raw instance to get the cHPI SNR from

cc @Eric89GXL

jjnurminen commented:
Hi @kambysese 
I'm not sure what you mean regarding the manuscript. I'll get on with the code review now and try to figure out what's going on here :)

jjnurminen commented:
IIRC the plan originally suggested by @Eric89GXL  was:

- the user will supply the evoked data
- the method will take the corresponding (unprocessed) raw data, epoch it and compute the cHPI SNR for each epoch
- weights according to the computed cHPI SNR will then be applied to the evoked data

Looking at the current version of the code, I cannot figure out the point where the unprocessed raw data is supposed to come in. @kambysese can you clarify?

larsoner commented:
> IIRC the plan originally suggested

Yeah that's about right. I agree the current code does not supply the original / unfiltered data, but it needs to. It could use `epochs.events` to get the corresponding time windows, though we'll have to be careful about decimation.

ktavabi commented:
> I cannot figure out the point where the unprocessed raw data is supposed to come in.

If we use `epochs.events` to determine the buffer lengths for cHPI SNR then the acquired raw data will have to be appropriately loaded using `_raw_LRFCP` windowed according and passed to `_chpi_snr_epochs`.  I suspect it can still all be done in the `save_epochs` block if @Eric89GXL are ok with that?

larsoner commented:
Yeah doing it within `save_epochs` should be fine, although we should only do it if chpi weighting is enabled.

ktavabi committed:
FIX: unfiltered epoch data for cHPI snr

Epoching raw and passing to chpi_snr_epochs for computing weights from unfiltered data.

ktavabi commented on mnefun/_mnefun.py:
```diff
@@ -1419,7 +1419,21 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
+                                               run_indices[si])
+                    raw = [Raw(fname, preload=False, allow_maxshield=True)
+                           for fname in raw_names]
+                    raw_unfil = concatenate_raws(raw)
```
~~@Eric89GXL `_raw_LRFCP` didn't really behave here in terms of filtering and resampling. Wondering if similar reasons prevented usage in ln 1331 of `save_epochs`?~~

ktavabi commented on mnefun/_mnefun.py:
```diff
@@ -1419,7 +1419,21 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
+                                       proj=None, preload=True,
+                                       decim=decim[si], picks=e.picks,
+                                       on_missing='warning')
+                    # Use unfiltered epochs for weights
```
@jjnurminen I am pretty sure that `eps_unfil` is what we want in terms of unfiltered raw data epoched as per user defined conditions that is subsequently passed to `_chpi_snr_epochs`.
@Eric89GXL Please vet this one in terms of sampling/filtering, I am pretty confident it's working as expected but I would appreciate a second opinion.

jjnurminen replied:
What we want is unprocessed data, preferably straight from the MEG. For example, lowpass filtering could remove the cHPI signals. I'm not sure what @Eric89GXL means by decimation above; if the raw data is downsampled before feeding it to chpi_snr, the new sampling freq must be high enough to capture the cHPI signals.

larsoner replied:
> I'm not sure what @Eric89GXL means by decimation above

I mean that if you take `epochs.events` from a `Epochs` that have been decimated, the event sample numbers `events[:, 0]` will be incorrect (off by a factor of `decim`). I agree you shoudl not decimate these epochs, as it will alias. Technically it's still possible to extract the cHPI frequencies at that point (I've used this trick in testing to speed things up before), but it decreases SNR (due to aliasing mixing) and is a huge pain, so we don't want to decimate here.

jjnurminen commented:
Ok, I see. Why would you have decimated data available, though? Everything
in there would be ruined by aliasing, right? OTOH, in properly downsampled
data (antialias filter followed by decimate) the HPI signals would be gone
(or hugely suppressed by the antialias filter).



On 9 Feb 2017 4:14 p.m., "Eric Larson" <notifications@github.com> wrote:

> *@Eric89GXL* commented on this pull request.
> ------------------------------
>
> In mnefun/_mnefun.py <https://github.com/jjnurminen/mnefun/pull/2>:
>
> > @@ -1419,7 +1419,21 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
>              evokeds = list()
>              for name in names:
>                  if p.average_weighting == 'chpi':
> -                    weights = _chpi_snr_epochs(e[name])
> +                    # read in raw unfiltered files for cHPI weighting evokeds
> +                    raw_names = get_raw_fnames(p, subj, 'raw', False, True,
> +                                               run_indices[si])
> +                    raw = [Raw(fname, preload=False, allow_maxshield=True)
> +                           for fname in raw_names]
> +                    raw_unfil = concatenate_raws(raw)
> +                    eps_unfil = Epochs(raw_unfil, events=e[name].events,
> +                                       event_id=None, tmin=p.tmin,
> +                                       tmax=p.tmax, baseline=_get_baseline(p),
> +                                       reject=None, flat=None,
> +                                       proj=None, preload=True,
> +                                       decim=decim[si], picks=e.picks,
> +                                       on_missing='warning')
> +                    # Use unfiltered epochs for weights
>
> I'm not sure what @Eric89GXL <https://github.com/Eric89GXL> means by
> decimation above
>
> I mean that if you take epochs.events from a Epochs that have been
> decimated, the event sample numbers events[:, 0] will be incorrect (off
> by a factor of decim). I agree you shoudl not decimate these epochs, as
> it will alias. Technically it's still possible to extract the cHPI
> frequencies at that point (I've used this trick in testing to speed things
> up before), but it decreases SNR (due to aliasing mixing) and is a huge
> pain, so we don't want to decimate here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jjnurminen/mnefun/pull/2>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIGROcYq4PPf-DME5DsKaGpGlo-SAZnGks5rax88gaJpZM4LRTvF>
> .
>

larsoner commented:
We typically low-pass and decimate, so the epochs object created in `save_epochs` is probably decimated
